Dsl2 conversion (#82)
seqeralabs/nf-training-public: asciidocs/intro.adoc

process convertToUpper { <13>

    input:  <14>
    file y <15>

    output: <16>
    stdout <17>

    """ <18>
    cat $y | tr '[a-z]' '[A-Z]'  <19>
    """ <20>
} <21>

workflow{ <22>
    letters_ch = splitLetters(greeting_ch) <23>
    results_ch = convertToUpper(letters_ch.flatten()) <24>
    results_ch.view{ it } <25>
} <26>
----
<1> The code begins with a shebang which declares Nextflow as the interepreter.
<2> Declares a parameter `greeting` that is initialized with the value 'Hello world!'.

> <3> Initalises a `channel` labelled `greeting_ch`, which contains the value from `params.greeting`. Channels are the necessary input type for processes in Nextflow.

Reviewer: ```suggestion
<3> Initializes a `channel` labelled `greeting_ch`, which contains the value from `params.greeting`. Channels are the necessary input type for processes in Nextflow.
```

<4> Begins the first process defined as `splitLetters`.
<5> Input declaration for the `splitLetters` process. Inputs can be values: `val`, files: `file`, paths `path` or other qualifiers (https://www.nextflow.io/docs/latest/process.html#inputs[See here])
<6> Tells the `process` to expect an input value (`val`), that we assign to the variable 'x'. 
<7> Output declaration for the `splitLetters` process.
<8> Tells the process to expect an output file/s (`file`) containing 'chunk_*' as output from the script and send the output as a channel. 

> <9> Triple quote marks intiate the code block to execute in this `process`.

Reviewer: ```suggestion
<9> Three double quote marks initiate the code block to execute in this `process`.
```

<10> Code to execute, printing the `input` value x (called using the dollar [$] symbol prefix), and splitting the string into chunks of 6 characters long ("Hello " and "world!") and saving to files: chunk_aa and chunk_ab.

> <11> Triple quote marks the end of the code block.

Reviewer: ```suggestion
<11> Three double quote marks end the code block.
```

<12> End of first process block.
<13> Begin second process defined as `convertToUpper`.
<14> Input declaration for the `convertToUpper` `process`.

> <15> Tells the `process` to expect `input` file/s (`file`; e.g. chunk_aa and chunk_ab), that we assign to the variable 'y'. 

Reviewer: ```suggestion
<15> Tells the `process` to expect an `input` file/s (`file`; e.g. chunk_aa and chunk_ab), that we assign to the variable 'y'. 
```

<16> Output declaration for the `convertToUpper` process.
<17> Tells the process to expect output as standard output (stdout) and direct this output as a channel.

> <18> Triple quote marks intiate the code block to execute in this `process`.

Reviewer: ```suggestion
<18> Triple quote marks initiate the code block to execute in this `process`.
```

> <18> Triple quote marks initiate the code block to execute in this `process`.

Reviewer: ```suggestion
<18> Three double quote marks initiate the code block to execute in this `process`.
```

<19> Script to read files (cat) using the '$y' input variable, then pipe to uppercase conversion, outputting to standard output.

> <20> Triple quote marks the end of the code block.

Reviewer: ```suggestion
<20> Three double quote marks end the code block.
```

* *6-7*: Followed by the printed result string from stdout.

TIP: The hexadecimal numbers, like `8a/537930`, identify the unique process
execution. These numbers are also the prefix of the directories where each
process is executed. You can inspect the files produced by changing to the directory
`$PWD/work` and use these numbers to find the process-specific
execution path.

IMPORTANT: The second process runs twice, executing in two different work directories 
for each input file. Therefore, in the previous example the work directory [9f/1dd42a] 
represents just one of the two directories that were processed. To print all the 

> relevent paths to screen, use the `-ansi-log` flag (e.g. `nextflow run hello.nf -ansi-log false`).

Reviewer: ```suggestion
relevant paths to screen, use the `-ansi-log` flag (e.g. `nextflow run hello.nf -ansi-log false`).
```
